ethereum/core/src/main/java/org/hyperledger/besu/ethereum/transaction/BlockSimulator.java

```diff
@@ -288,94 +320,63 @@ protected void applyStateOverrides(
-        .buildBlockHeader();
-  }
-
-  /**
-   * Creates the final block header after applying state changes and transaction processing.
-   *
-   * @param blockHeader The original block header.
-   * @param ws The MutableWorldState after applying state overrides.
-   * @param transactions The list of transactions in the block.
-   * @param blockOverrides The BlockOverrides to apply.
-   * @param receipts The list of transaction receipts.
-   * @param currentGasUsed The total gas used in the block.
-   * @return The final block header.
-   */
-  private BlockHeader createFinalBlockHeader(
-      final BlockHeader blockHeader,
-      final MutableWorldState ws,
-      final List<Transaction> transactions,
       final BlockOverrides blockOverrides,
-      final List<TransactionReceipt> receipts,
-      final long currentGasUsed) {
-
-    return BlockHeaderBuilder.createDefault()
-        .populateFrom(blockHeader)
-        .ommersHash(BodyValidation.ommersHash(List.of()))
-        .stateRoot(blockOverrides.getStateRoot().orElse(ws.rootHash()))
-        .transactionsRoot(BodyValidation.transactionsRoot(transactions))
-        .receiptsRoot(BodyValidation.receiptsRoot(receipts))
-        .logsBloom(BodyValidation.logsBloom(receipts))
-        .gasUsed(currentGasUsed)
-        .withdrawalsRoot(null)
-        .requestsHash(null)
-        .mixHash(blockOverrides.getMixHashOrPrevRandao().orElse(Hash.EMPTY))
-        .extraData(blockOverrides.getExtraData().orElse(Bytes.EMPTY))
-        .blockHeaderFunctions(new SimulatorBlockHeaderFunctions(blockOverrides))
+      final boolean shouldValidate) {
+    long timestamp = blockOverrides.getTimestamp().orElseThrow();
+    long blockNumber = blockOverrides.getBlockNumber().orElseThrow();
+
+    BlockHeaderBuilder builder =
+        BlockHeaderBuilder.createDefault()
+            .parentHash(header.getHash())
+            .timestamp(timestamp)
+            .number(blockNumber)
+            .coinbase(
+                blockOverrides
+                    .getFeeRecipient()
+                    .orElseGet(() -> miningConfiguration.getCoinbase().orElseThrow()))
+            .difficulty(
+                blockOverrides.getDifficulty().map(Difficulty::of).orElseGet(header::getDifficulty))
+            .gasLimit(
+                blockOverrides
+                    .getGasLimit()
+                    .orElseGet(() -> getNextGasLimit(newProtocolSpec, header, blockNumber)))
+            .baseFee(
+                blockOverrides
+                    .getBaseFeePerGas()
+                    .orElseGet(
+                        () ->
+                            shouldValidate
+                                ? getNextBaseFee(newProtocolSpec, header, blockNumber)
+                                : Wei.ZERO))
+            .extraData(blockOverrides.getExtraData().orElse(Bytes.EMPTY))
+            .parentBeaconBlockRoot(Bytes32.ZERO)
+            .prevRandao(blockOverrides.getMixHashOrPrevRandao().orElse(Bytes32.ZERO));
+
```

Comment:
what about moving the functionality of overriding an existing block in the `BlockHeaderBuilder` class for general use, for example `BlockHeaderBuilder::copyWithOverrides`?

Reply:
This is a nice suggestion, however in this case, it is not only a copy with overrides, the block simulator is assembling the next block base for the simulation, so it needs data from other sources such as `getNextGasLimit`, `getNextBaseFee`, `miningConfiguration`, etc.